### ISSUE TITLE
Unknown builtin op: torchvision::roi_align.

### ISSUE THREAD

**Issue by taqta** [pytorch/vision#7594]
### 🐛 Describe the bug

Hi, I met an error when using torch.jit.load()
my code is as below:

      video_model = build_model(cfg).cpu()
      ckp_state = torch.load('/home/dsphaoyang/yolo_slowfast/SLOWFAST_32x2_R101_50_50.pkl',
                                                                                                                                map_location=torch.device('cpu'))
      ckp_state = ckp_state['model_state']
      video_model.load_state_dict(ckp_state)
      video_model=video_model.eval()
      store = torch.jit.trace(video_model,(input[0],input[1],box))
      
      store.save('./slowfast101_trace_ver3.torchscript')
      t=torch.jit.load('slowfast101_trace_ver3.torchscript')

and the error is:

      Unknown builtin op: torchvision::roi_align.
      Could not find any similar ops to torchvision::roi_align. This op may not exist or may not be currently supported in TorchScript.
      :
      File "code/__torch__/detectron2/layers/roi_align.py", line 10
          argument_2: Tensor) -> Tuple[Tensor, Tensor]:
          boxes = torch.to(bboxes, 6)
          input = ops.torchvision.roi_align(argument_2, boxes, 0.0625, 7, 7, 0, False)
                  ~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
          return (input, boxes)



### Versions

PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Clang version: Could not collect
CMake version: version 3.26.3
Libc version: glibc-2.17

Python version: 3.9.0 (default, Nov 15 2020, 14:28:56)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.11.1.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: 10.2.89
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: 
GPU 0: Tesla K80
GPU 1: Tesla K80

Nvidia driver version: 430.26
cuDNN version: /data/software/cuda/10.1/targets/x86_64-linux/lib/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                32
On-line CPU(s) list:   0-31
Thread(s) per core:    2
Core(s) per socket:    8
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 79
Model name:            Intel(R) Xeon(R) CPU E5-2620 v4 @ 2.10GHz
Stepping:              1
CPU MHz:               1200.347
CPU max MHz:           3000.0000
CPU min MHz:           1200.0000
BogoMIPS:              4200.03
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              20480K
NUMA node0 CPU(s):     0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30
NUMA node1 CPU(s):     1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cdp_l3 invpcid_single intel_pt tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a rdseed adx smap xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] pytorchvideo==0.1.5
[pip3] torch==2.0.1
[pip3] torchaudio==2.0.2
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] pytorchvideo              0.1.5                    pypi_0    pypi
[conda] torch                     2.0.1                    pypi_0    pypi
[conda] torchaudio                2.0.2                    pypi_0    pypi
[conda] torchvision               0.15.2                   pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

**Comment by NicolasHug** [comment]
What version of `detectron2` are you using?

I suspect this is a version issue. Looking at the original repo where you took the model from, the requirements are for torchvision 0.9 https://github.com/wufan-tb/yolo_slowfast/blob/master/requirements.txt

**Comment by taqta** [comment]
> What version of `detectron2` are you using?
> 
> I suspect this is a version issue. Looking at the original repo where you took the model from, the requirements are for torchvision 0.9 https://github.com/wufan-tb/yolo_slowfast/blob/master/requirements.txt

Thanks a lot for your reply! But when I tried to run the script with in the environment with torchvision 0.9.1, I get the same result. So, I guess it may not be the version issue. And I also tried the code below:

             store = torch.jit.script(video_model,(input[0],input[1],box))
             torch.jit.save(store,'./slowfast101_trace_ver5.torchscript')

I got the an different error in torch.jit.save:
              RuntimeError: strides() called on an undefined Tensor

**Comment by taqta** [comment]
> What version of `detectron2` are you using?
> 
> I suspect this is a version issue. Looking at the original repo where you took the model from, the requirements are for torchvision 0.9 https://github.com/wufan-tb/yolo_slowfast/blob/master/requirements.txt

Detection2 is seldom used in this project, so it may not cause this problem.

**Comment by NicolasHug** [comment]
All I can tell you is that `torchvision.ops.roi_align` supports torchscript (we have various tests ensure that).

Now I do really know for sure what the cause of the issue is, but I suspect it has to do with versions, or with outdated naming. 

> ops.torchvision.roi_align

Seeing this coming from `detectron2` looks a bit fishy to me. There is no `torchvision.roi_align`, it's `torchvision.ops.roi_align`. So there might be something weird going on in `detectron2` or elsewhere in the stack, but I honestly doubt the issue comes from torchvision directly.

**Comment by taqta** [comment]
Thank you for your reply. 
the original code in detection2.layer.roi_align is:

          from torch import nn
          from torchvision.ops import roi_align
          class ROIAlign(nn.Module):
              def __init__(self, output_size, spatial_scale, sampling_ratio, aligned=True):
                  super().__init__()
                  self.output_size = output_size
                  self.spatial_scale = spatial_scale
                  self.sampling_ratio = sampling_ratio
                  self.aligned = aligned
          
                  from torchvision import __version__
          
                  version = tuple(int(x) for x in __version__.split(".")[:2])
                  # https://github.com/pytorch/vision/pull/2438
                  assert version >= (0, 7), "Require torchvision >= 0.7"
          
              def forward(self, input, rois):
                  """
                  Args:
                      input: NCHW images
                      rois: Bx5 boxes. First column is the index into N. The other 4 columns are xyxy.
                  """
                  assert rois.dim() == 2 and rois.size(1) == 5
                  if input.is_quantized:
                      input = input.dequantize()
                  return roi_align(
                      input,
                      rois.to(dtype=input.dtype),
                      self.output_size,
                      self.spatial_scale,
                      self.sampling_ratio,
                      self.aligned,
                  )

**Comment by NicolasHug** [comment]
That doesn't correspond to your trace:

```
  File "code/__torch__/detectron2/layers/roi_align.py", line 10
      argument_2: Tensor) -> Tuple[Tensor, Tensor]:
      boxes = torch.to(bboxes, 6)
      input = ops.torchvision.roi_align(argument_2, boxes, 0.0625, 7, 7, 0, False)
              ~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
      return (input, boxes)
```

**Comment by taqta** [comment]
I replace detection.layers.roi_align with torchvision.ops.RoIAlign in my code. And I change  torch's version to 1.8.0,torchvision to 0.9.0. But I get the similar error after I run the code below:

          video_model = build_model(cfg).cpu()
          ckp_state = torch.load('/home/dsphaoyang/yolo_slowfast/SLOWFAST_32x2_R101_50_50.pkl',
                                 map_location=torch.device('cpu'))
          ckp_state = ckp_state['model_state']
          video_model.load_state_dict(ckp_state)
          video_model=video_model.eval()
          store = torch.jit.trace(video_model,(input[0],input[1],box))
          torch.jit.save(store,'./slowfast101_trace_ver8.pt')
          x=torch.jit.load('./slowfast101_trace_ver8.pt')

the error is:

    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "/home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/jit/_serialization.py", line 161, in load
        cpp_module = torch._C.import_ir_module(cu, str(f), map_location, _extra_files)
    RuntimeError: 
    Unknown builtin op: torchvision::roi_align.
    Could not find any similar ops to torchvision::roi_align. This op may not exist or may not be currently supported in TorchScript.
    :
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torchvision/ops/roi_align.py(53): roi_align
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torchvision/ops/roi_align.py(76): forward
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/nn/modules/module.py(860): _slow_forward
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/nn/modules/module.py(887): _call_impl
    /home/dsphaoyang/yolo_slowfast/slowfast/models/head_helper.py(148): forward
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/nn/modules/module.py(860): _slow_forward
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/nn/modules/module.py(887): _call_impl
    /home/dsphaoyang/yolo_slowfast/slowfast/models/video_model_builder.py(446): forward
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/nn/modules/module.py(860): _slow_forward
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/nn/modules/module.py(887): _call_impl
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/jit/_trace.py(934): trace_module
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/jit/_trace.py(733): trace
    /home/dsphaoyang/yolo_slowfast/get_trace.py(69): <module>
    Serialized   File "code/__torch__/torchvision/ops/roi_align.py", line 9
        boxes: Tensor,
        input: Tensor) -> Tensor:
        input0 = ops.torchvision.roi_align(input, boxes, 0.0625, 7, 7, 0, False)
                 ~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
        return input0

**Comment by taqta** [comment]
I tried the code below:

      import torch
      import torchvision
      
      roi_align=torchvision.ops.RoIAlign(output_size=[7,7],spatial_scale=0.0625,sampling_ratio=0,aligned=False)
      bbox=torch.rand([2,5])
      input=torch.rand([1,2048,40,72])
      
      trace=torch.jit.trace(roi_align,(input,bbox))
      torch.jit.save(trace,'try.pt')
      torch.jit.load('try.pt')

and I meet the same error:


     Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "/home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/jit/_serialization.py", line 161, in load
    cpp_module = torch._C.import_ir_module(cu, str(f), map_location, _extra_files)
    RuntimeError: 
    Unknown builtin op: torchvision::roi_align.
    Could not find any similar ops to torchvision::roi_align. This op may not exist or may not be currently supported in TorchScript.
    :
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torchvision/ops/roi_align.py(53): roi_align
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torchvision/ops/roi_align.py(76): forward
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/nn/modules/module.py(860): _slow_forward
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/nn/modules/module.py(887): _call_impl
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/jit/_trace.py(934): trace_module
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/jit/_trace.py(733): trace
    /home/dsphaoyang/yolo_slowfast/try_roi.py(8): <module>
    Serialized   File "code/__torch__/torchvision/ops/roi_align.py", line 9
        input: Tensor,
        boxes: Tensor) -> Tensor:
        _0 = ops.torchvision.roi_align(input, boxes, 0.0625, 7, 7, 0, False)
             ~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
        return _0

by the way, If I run the same code in script, it won't report error. Only if I run the code in terminal or python console, would it report the above error. 





> That doesn't correspond to your trace:
> 
> ```
>   File "code/__torch__/detectron2/layers/roi_align.py", line 10
>       argument_2: Tensor) -> Tuple[Tensor, Tensor]:
>       boxes = torch.to(bboxes, 6)
>       input = ops.torchvision.roi_align(argument_2, boxes, 0.0625, 7, 7, 0, False)
>               ~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
>       return (input, boxes)
> ```

**Comment by pmeier** [comment]
I cannot reproduce the issue with the snippet you posted. Neither on the current nightlies nor on the latest release.

> by the way, If I run the same code in script, it won't report error. Only if I run the code in terminal or python console, would it report the above error.

In that case, your setup is incorrect somewhere. Closing this, because it sounds like an env issue. If you can reproduce in a clean env by running a script, feel free to re-open.

---

Probably unrelated, but your env is fishy in multiple other ways:

- You have Python `3.9.0` released ~2.5 years ago. You probably want the current `3.9.16` release, which is only half a year old at this point.
- Your Nvidia driver version `430.26` is really outdated. With that you cannot run CUDA 10 let alone the CUDA 11.7 versions that you have installed in your env. For that you need at least `450.80.02 ` ([Table 2](https://docs.nvidia.com/cuda/archive/11.8.0/cuda-toolkit-release-notes/index.html))

**Comment by taqta** [comment]
Thanks greatly for your reply! I have found out what the problem is. 

When I run the code below:

    import torch 
    x=torch.jit.load('try.pt')
    #where the try.pt is generated by the code I have mentioned.

then, the error is report:

    Traceback (most recent call last):
      File "/home/dsphaoyang/.pycharm_helpers/pydev/pydevconsole.py", line 364, in runcode
        coro = func()
      File "<input>", line 1, in <module>
      File "/home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/jit/_serialization.py", line 161, in load
        cpp_module = torch._C.import_ir_module(cu, str(f), map_location, _extra_files)
    RuntimeError: 
    Unknown builtin op: torchvision::roi_align.
    Could not find any similar ops to torchvision::roi_align. This op may not exist or may not be currently supported in TorchScript.
    :
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torchvision/ops/roi_align.py(53): roi_align
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torchvision/ops/roi_align.py(76): forward
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/nn/modules/module.py(860): _slow_forward
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/nn/modules/module.py(887): _call_impl
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/jit/_trace.py(934): trace_module
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/site-packages/torch/jit/_trace.py(733): trace
    <input>(8): <module>
    /home/dsphaoyang/.pycharm_helpers/pydev/pydevconsole.py(364): runcode
    /home/dsphaoyang/anaconda3/envs/yolo_slowfast/lib/python3.9/code.py(74): runsource
    /home/dsphaoyang/.pycharm_helpers/pydev/_pydev_bundle/pydev_console_types.py(35): run
    /home/dsphaoyang/.pycharm_helpers/pydev/pydevconsole.py(93): do_add_exec
    /home/dsphaoyang/.pycharm_helpers/pydev/_pydev_bundle/pydev_code_executor.py(109): add_exec
    /home/dsphaoyang/.pycharm_helpers/pydev/pydevconsole.py(287): process_exec_queue
    /home/dsphaoyang/.pycharm_helpers/pydev/pydevconsole.py(498): start_client
    /home/dsphaoyang/.pycharm_helpers/pydev/pydevconsole.py(570): <module>
    Serialized   File "code/__torch__/torchvision/ops/roi_align.py", line 9
        input: Tensor,
        boxes: Tensor) -> Tensor:
        _0 = ops.torchvision.roi_align(input, boxes, 0.0625, 7, 7, 0, False)
             ~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
        return _0

But when I run the code below:

    import torch
    import torchvision
    x=torch.jit.load('try.pt')

No error is reported

it has nothing to do with python console or terminal. The reason I met the error above when I use terminal or console is that I   did not import torchvision for convenience. And the same situation happened in serveral torch version.
